Typed iterator in QASM3 export of for loops

arthurostrauss opened the pull request:
### Summary

Fixes an issue in the QASM3 exporter where the type of the loop variable in `for` loops was not being explicitly included in the generated code. This could lead to invalid OpenQASM 3 code. This solves #13725.

### Details and comments

The exporter now explicitly adds the type (int) to the `for` loop declaration. A deprecation warning is raised if the type is not specified. The tests have been updated to reflect this change.

arthurostrauss committed:
Adds explicit type to QASM3 for loops

Adds an explicit type to QASM3 for loops, defaulting to `int`
if no type is specified. This change also introduces a
deprecation warning for typeless for loops which will be
removed in a future release. The exporter is updated to always
include the int type, and tests are updated to reflect the
new syntax.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into fix_qasm3_for_loop

jakelishman commented on qiskit/qasm3/ast.py:
```diff
@@ -697,17 +698,26 @@ class ForLoopStatement(Statement):
+                "Specify the type parameter explicitly.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
```
We don't actually need to do this because `qasm3.ast` isn't a public module nor any of its contents, but it doesn't really hurt.

jakelishman commented on qiskit/qasm3/ast.py:
```diff
@@ -697,17 +698,26 @@ class ForLoopStatement(Statement):
         self.indexset = indexset
         self.parameter = parameter
         self.body = body
+        self.type = type_ if type_ is not None else IntType()
```
The AST constructor shouldn't really be making assumptions about types.  It's either got a type, or it hasn't.  If you very much want to try and provide a deprecation warning, then we'd just maintain the `None` here, and have the printer not print a type if it's `None`, rather than defaulting to something arbitrary.

Really though, it's fine just to immediately require that this is a non-optional - the class isn't public and shouldn't be constructed by anything other than Qiskit internals.

arthurostrauss replied:
But not printing a type in OpenQASM3 is invalid now isn't it?

arthurostrauss replied:
Furthermore, this is just to comply with the current status of what can be expressed with a ForLoop (note that in my other PR when I introduce the expr.Range object, we could indeed think about relaxing the type of the start stop and steps to make it more flexible with current Qiskit available types in latest versions. But this points back to my suggestion of adding expr.Array() as well, we might want to loop over real time variables in general.

jakelishman commented on qiskit/qasm3/exporter.py:
```diff
@@ -1229,7 +1229,8 @@ def build_for_loop(self, instruction: CircuitInstruction) -> ast.ForLoopStatemen
                     ) from None
             body_ast = ast.ProgramBlock(self.build_current_scope())
-        return ast.ForLoopStatement(indexset_ast, loop_parameter_ast, body_ast)
+            type_ = ast.IntType()
```
The same OQ3 language change that introduced the requirement to type the loop variable also made it possible to have loop variables that are not integer typed, so from an AST perspective, there's no particular reason that this has to be `IntType`.  That said, the Qiskit `ForLoopOp` is still _typed_ as being an `Iterable[int]`, so I guess we're ok for now.

arthurostrauss replied:
Yes this was done to comply with the current status of Qiskit ForLoop index set. @1ucian0 wanted this as a back port fix for earlier version.

jakelishman reviewed (commented):
Thanks for fixing this - it's sat for a long time since OQ3 changed the syntax of the language to require typing variables.

arthurostrauss committed:
Removed deprecation warning in AST for Type specification

The type in the ForLoopStatement AST can no longer be optional. Instead, the exporter complies with latest Qiskit specification by forcing the IntType. The deprecation warning upon not providing a type to the AST statement has hence been removed.

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into fix_qasm3_for_loop

arthurostrauss committed:
Linting (removed unused import)

1ucian0 commented:
As this PR is blocking https://github.com/Qiskit/qiskit/pull/14618 , should it be tagged for 2.3 milestone?

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into fix_qasm3_for_loop

arthurostrauss committed:
Merge branch 'main' into fix_qasm3_for_loop

arthurostrauss committed:
Merge branch 'main' into fix_qasm3_for_loop

arthurostrauss committed:
Merge remote-tracking branch 'Qiskit/main' into fix_qasm3_for_loop

arthurostrauss committed:
Added upgrade note

arthurostrauss committed:
Updated upgrade name

arthurostrauss committed:
Additional fix

1ucian0 commented on releasenotes/notes/fix_qasm3_export_for_loop-23429092276a53fa.yaml:
```diff
@@ -0,0 +1,13 @@
+  - |
+    The OpenQASM 3 exporter (:mod:`qiskit.qasm3`) will now correctly export for loops with an explicit type.
+    By default, the type is now `int`, but it can be overridden by passing a `type` argument to the ForLoopStatement constructor in
+    the qasm3.ast module.
```
```suggestion
    The OpenQASM 3 exporter (:mod:`qiskit.qasm3`) now includes the type for the iterator in for loops. Fixes `#13725 <https://github.com/Qiskit/qiskit/issues/13725>`__.

upgrade_qasm:
  - |
    Previous OpenQASM 3 code generated by :mod:`qiskit.qasm3` did not include the type in the iterator variable. The exporter now includes an explicit type, following the OpenQASM 3.0 specification.
```

1ucian0 committed:
Update releasenotes/notes/fix_qasm3_export_for_loop-23429092276a53fa.yaml

1ucian0 commented:
Thank for the fix!
